src/analysis_and_optimization/Partial_evaluator.ml

```diff
@@ -6,11 +6,12 @@ open Middle
 
 exception Rejected of Location_span.t * string
 
-let is_int i Expr.Fixed.{pattern; _} =
+let rec is_int i Expr.Fixed.{pattern; _} =
   let nums = List.map ~f:(fun s -> string_of_int i ^ s) [""; "."; ".0"] in
```

Comment:
Agree, changing to 

```ocaml
let rec is_int query Expr.Fixed.{pattern; _} =
  match pattern with
  | Lit (Int, i) | Lit (Real, i) -> float_of_string i = float_of_int query
  | Promotion (e, _, _) -> is_int query e
  | _ -> false
```

As for `eval_expr`, I'm not sure if having cases for Promote(Lit (...), ...) is that useful, but we could do it (at least for `Promote (Lit (Int, _), UReal, _)`)

Reply:
A case for `Promotion (ArrayExpr _, _)` might be worthwhile. Anyway, debug data generation basically assumes that the partial evaluator is able to evaluate any constant expression into a literal (which is not really possible if there are user-defined functions).

Reply:
The changes in #1102 actually make it so you should never see `Promotion(ArrayExpr ...` - The logic for promotions was updated so promotion happens at the smallest level possible, to avoid generating two copies in the C++ 

e.g, `promote<double>(std::vector<int>{1,2,3})` gets inside-outed such that it is actually `std::vector<double>{promote<double>(1), ...})`. This was done on a case-by-case bases in the code generator before, and is now represented in the AST. The extra cases in codegen ensure that `Promote (<integer scalar value>, UReal, _)` are no-ops and just print the integer scalar value.

For debug data, I added a bit of code to `unwrap_num_exn` which causes it to strip off promotions in anticipating of this problem. Changing the partial evaluator directly causes a lot of the generated C++ to change for the optimization outputs, but I think we could do that instead if we wanted

Reply:
This again ignores MIR optimizations.
Example:
```stan
model {
  int abc[2] = {1, 2};
  real def[2] = abc;
  target += sum(def);
}
```
After `expression_propagation` but before `lazy_code_motion` (which is how the partial evaluator sees it) the optimized MIR is
```
log_prob {
  {
    data array[int, 2] abc;
    data array[real, 2] def;
    target += sum(promote(FnMakeArray__(1, 2), real));
  }
}
```